executor/parallel_apply.go

```diff
@@ -0,0 +1,392 @@
+	}
+
+	if e.useCache {
+		if e.cache, err = newApplyCache(e.ctx); err != nil {
+			return err
+		}
+		e.cache.GetMemTracker().AttachTo(e.memTracker)
+	}
+	return nil
+}
+
+// Next implements the Executor interface.
+func (e *ParallelNestedLoopApplyExec) Next(ctx context.Context, req *chunk.Chunk) (err error) {
+	if !e.started {
+		e.workerWg.Add(1)
+		go e.outerWorker(ctx)
+		for i := 0; i < e.concurrency; i++ {
+			e.workerWg.Add(1)
+			atomic.AddInt32(&e.numWorkers, 1)
+			go e.innerWorker(ctx, i)
+		}
+		e.notifyWg.Add(1)
+		go e.notifyWorker(ctx)
+		e.started = true
+	}
+	result := <-e.resultChkCh
+	if result.err != nil {
+		return result.err
+	}
+	req.SwapColumns(result.chk)
+	e.freeChkCh <- result.chk
+	return nil
+}
+
+// Close implements the Executor interface.
+func (e *ParallelNestedLoopApplyExec) Close() error {
+	e.memTracker = nil
+	err := e.outerExec.Close()
+	if e.started {
+		close(e.exit)
+		e.started = false
+		e.notifyWg.Wait()
+	}
+	if e.runtimeStats != nil {
+		if e.useCache {
+			var hitRatio float64
+			if e.cacheAccessCounter > 0 {
+				hitRatio = float64(e.cacheHitCounter) / float64(e.cacheAccessCounter)
+			}
+			e.runtimeStats.SetCacheInfo(true, hitRatio)
+		} else {
+			e.runtimeStats.SetCacheInfo(false, 0)
+		}
+	}
+	return err
+}
+
+// notifyWorker waits for all inner/outer-workers finishing and then put an empty
+// chunk into the resultCh to notify the upper executor there is no more data.
+func (e *ParallelNestedLoopApplyExec) notifyWorker(ctx context.Context) {
+	e.workerWg.Wait()
+	e.putResult(nil, nil)
+	e.notifyWg.Done()
+}
+
+func (e *ParallelNestedLoopApplyExec) outerWorker(ctx context.Context) {
+	defer e.handleWorkerPanic(ctx)
+	var selected []bool
+	var err error
+	for {
+		chk := newFirstChunk(e.outerExec)
+		if err := Next(ctx, e.outerExec, chk); err != nil {
+			e.putResult(nil, err)
+			return
+		}
+		if chk.NumRows() == 0 {
+			close(e.outerRowCh)
+			return
+		}
+		e.outerList.Add(chk)
+		outerIter := chunk.NewIterator4Chunk(chk)
+		selected, err = expression.VectorizedFilter(e.ctx, e.outerFilter, outerIter, selected)
+		if err != nil {
+			e.putResult(nil, err)
+			return
+		}
+		for i := 0; i < chk.NumRows(); i++ {
+			row := chk.GetRow(i)
+			select {
+			case e.outerRowCh <- outerRow{&row, selected[i]}:
+			case <-e.exit:
+				return
+			}
+		}
+	}
+}
+
+func (e *ParallelNestedLoopApplyExec) innerWorker(ctx context.Context, id int) {
+	defer e.handleWorkerPanic(ctx)
+	for {
+		var chk *chunk.Chunk
+		select {
+		case chk = <-e.freeChkCh:
+		case <-e.exit:
+			return
+		}
+		err := e.fillInnerChunk(ctx, id, chk)
+		if err == nil && chk.NumRows() == 0 { // no more data, this goroutine can exit
+			for {
+				numWorkers := atomic.LoadInt32(&e.numWorkers) // decrease the numWorkers counter
+				if atomic.CompareAndSwapInt32(&e.numWorkers, numWorkers, numWorkers-1) {
+					if numWorkers == 1 { // it's the last one, deliver an empty chunk to notify the upper executor
```

Comment:
atomic.addInt32(&e.numWorkers, -1) is better.

Reply:
I remove this logic now since there is another notify goroutine to do this work. https://github.com/pingcap/tidb/pull/18044/commits/2eb56537016a483cb2e615e55701e194069e010a